### ISSUE TITLE
Use ES6 in hoxy, particularly generators.

### ISSUE THREAD

**Issue by greim** [greim/hoxy#38]
**PROS:**

Reduces dependence on callbacks. Writing code in a synchronous/suspended style would make it easier to reason about and refactor.

**CONS:**

Limits hoxy use to newer node platforms, namely iojs or node 0.11 / `--harmony`.

Cons might be mitigated by use of transpilers. Also by hoxy being a dev-focused tool rather than a production one, thus not locked to mission-critical systems and more under control of devs.

If/when iojs becomes stable/popular and/or Joyent releases node 0.12 with native generator support the cons will be minimized in my view.

**Comment by sholladay** [comment]
@greim It's awesome that this is done. I'm now trying to figure out how to use it. Am I correct in understanding that an ES6 module which `import`s hoxy 2.x (and sits in a package having hoxy installed as a typical package.json `dependency`) is to be run with `babel-node`?

I keep getting stuff like this...

````javascript
/Users/sholladay/Code/experiment/es6-proxy/node_modules/hoxy/node_modules/babel/node_modules/babel-core/lib/babel/polyfill.js:16
  throw new Error("only one instance of babel/polyfill is allowed");
        ^
Error: only one instance of babel/polyfill is allowed
    at Object.<anonymous> (/Users/sholladay/Code/experiment/es6-proxy/node_modules/hoxy/node_modules/babel/node_modules/babel-core/lib/babel/polyfill.js:16:9)
    at Module._compile (module.js:460:26)
    at Module._extensions..js (module.js:478:10)
    at Object.require.extensions.(anonymous function) [as .js] (/usr/local/lib/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:215:7)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/Users/sholladay/Code/experiment/es6-proxy/node_modules/hoxy/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node-polyfill.js:11:1)
    at Module._compile (module.js:460:26)
````

But using typical `node` doesn't work, either ...

````javascript
/Users/sholladay/Code/experiment/es6-proxy/node_modules/hoxy/lib/proxy.js:6
import http from 'http'
^^^^^^
SyntaxError: Unexpected reserved word
    at exports.runInThisContext (vm.js:73:16)
    at Module._compile (module.js:443:25)
    at Module._extensions..js (module.js:478:10)
    at Object.require.extensions.(anonymous function) [as .js] (/Users/sholladay/Code/sitecues/sitecues-proxy/node_modules/hoxy/node_modules/babel/node_modules/babel-core/lib/babel/api/register/node.js:215:7)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/Users/sholladay/Code/sitecues/sitecues-proxy/node_modules/hoxy/hoxy.js:9:13)
    at Module._compile (module.js:460:26)
````

There must be something basic I'm missing here.

**Comment by greim** [comment]
Hoxy being written in es6 is *supposed* to be invisible to the outside world. AKA you should be able to `npm install hoxy` into es5 projects, `var hoxy = require('hoxy')` from es5 modules, and run it using node v0.10 or higher without `--harmony` flags.

Obviously something isn't playing well with your environment. You definitely shouldn't need babel-node for Hoxy, but perhaps your babel-node is a different version from the one Hoxy depends on? I suspect something like [this](http://babeljs.io/docs/usage/runtime/) needs to be added to Hoxy.

**Comment by sholladay** [comment]
Right. Okay, we're on the same page. I only tried `babel-node` after reaching a dead-end with ES5 at:
````javascript
import http from 'http'
^^^^^^
SyntaxError: Unexpected reserved word
````

(last comment has full trace)

My limited understanding of how Babel works tells me that we either need that runtime like you mentioned or that hoxy needs to be transpiled at `prepublish` time when pushing up to npm with [a package script](https://docs.npmjs.com/misc/scripts).

**Comment by greim** [comment]
It will definitely be good to do some of these things that generally make the lib more self-contained, but if you could describe the minimum conditions to trigger this specific error, that would be helpful too. And maybe filed as a separate issue.

**Comment by greim** [comment]
Just published 2.1.1 which uses a prepublish script. Let me know whether that works.

**Comment by majdamarcin** [comment]
Unfortunately, it's not working for me:
```
λ node proxy.js
module.js:338
    throw err;
          ^
Error: Cannot find module './lib/proxy'
    at Function.Module._resolveFilename (module.js:336:15)
    at Function.Module._load (module.js:278:25)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (c:\xampp\htdocs\mms\node_modules\hoxy\index.js:6:13)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
```
In proxy.js is only one line :
```
var hoxy = require('hoxy');
```
I'm using node v0.12.0.

**Comment by greim** [comment]
@majdamarcin - Are you pulling directly from this repo, instead of from npm? If so, you'll need to run `npm run compile` before anything will work. If not, could you describe your setup so that I can try to replicate the issue?

**Comment by greim** [comment]
Hmm. There was an error in the .npmignore file which is fixed in v2.2.3. Let me know whether that fixes this issue.

**Comment by majdamarcin** [comment]
@greim I just installed version v2.2.3 and everything looks fine, thanks :clap:

**Comment by majdamarcin** [comment]
I'm sorry for disinformation, but I just tried to write simple script: 
```
var hoxy = require('hoxy');
var proxy = new hoxy.Proxy();

proxy.intercept({
    phase: 'response',
    url: '/api/v1/sample/',
    methdod: 'POST',
    mimeType: 'application/json',
    as: 'json'
}, function(req, resp) {
    console.log(resp.userProfile.roles);
});
proxy.listen(8001);
```
and unfortunately it's not working. I get an error:
```
*truncated*\node_modules\hoxy\lib\proxy.js:263
      _co2['default'].call(_this2, regeneratorRuntime.mark(function callee$3$0
                                   ^
ReferenceError: regeneratorRuntime is not defined
    at Server.<anonymous> (*truncated*\node_modules\hoxy\lib\proxy.js:263:36)
    at Server.emit (events.js:98:17)
    at HTTPParser.parser.onIncoming (http.js:2113:12)
    at HTTPParser.parserOnHeadersComplete [as onHeadersComplete] (http.js:122:23)
    at Socket.socket.ondata (http.js:1971:22)
    at TCP.onread (net.js:528:27)
```

I installed hoxy via npm, my configuration:
```
λ node -v
v0.10.36

λ cat package.json
{
  "name": "app",
  "version": "0.1.0",
  "devDependencies": {
    "bluebird": "^2.9.34",
    "chromedriver": "~2.8.0",
    "grunt": "~0.4.2",
    "grunt-contrib-clean": "~0.4.1",
    "grunt-contrib-compass": "~0.2.0",
    "grunt-contrib-concat": "~0.1.3",
    "grunt-contrib-copy": "~0.5.0",
    "grunt-contrib-jshint": "~0.3.0",
    "grunt-contrib-uglify": "~0.2.0",
    "grunt-contrib-watch": "~0.5.3",
    "grunt-contrib-yuidoc": "~0.4.0",
    "grunt-css": "~0.5.4",
    "grunt-ddescribe-iit": "0.0.6",
    "grunt-html2js": "~0.2.4",
    "grunt-hub": "~0.6.0",
    "grunt-karma": "~0.8.3",
    "grunt-protractor-runner": "^2.0.0",
    "grunt-remove-logging": "~0.1.1",
    "grunt-replace": "~0.4.2",
    "grunt-sass": "~0.4.0",
    "grunt-selenium-simple": "~0.1.1",
    "hoxy": "^2.2.3",
    "install": "^0.1.8",
    "jasmine-core": "^2.2.0",
    "jasmine-spec-reporter": "^1.1.2",
    "karma": "~0.12.19",
    "karma-chrome-launcher": "~0.1.4",
    "karma-firefox-launcher": "~0.1.3",
    "karma-html-reporter": "^0.2.6",
    "karma-jasmine": "^0.3.5",
    "karma-phantomjs-launcher": "~0.1.4",
    "karma-safari-launcher": "~0.1",
    "karma-verbose-reporter": "0.0.1",
    "moment": "^2.10.3",
    "npm": "^2.6.1",
    "phantomjs": "~1.9.7",
    "protractor": "^2.1.0",
    "protractor-html-screenshot-reporter": "0.0.19",
    "selenium": "~2.20.0",
    "underscore": "^1.8.3"
  }

**Comment by nerdbeere** [comment]
@majdamarcin I had the same issue while using browserify. 
You need the babel polyfill as described [in the docs](https://babeljs.io/docs/usage/polyfill/) and in [this stack overflow question](http://stackoverflow.com/questions/28976748/regeneratorruntime-is-not-defined).

Adding `require('babel/polyfill');` fixed it for me.

**Comment by majdamarcin** [comment]
@nerdbeere  thank you very much!
```
npm install babel --save-dev
```
and 
```
require('babel/polyfill');
```
resolved my issues.

**Comment by greim** [comment]
Thanks for the workaround @nerdbeere. I believe this is a bug in Hoxy and I'll have a fix ready to go soon. This is the first time I've published an npm module using babel so thanks for bearing with me.

**Comment by greim** [comment]
Fix published in 2.2.4.
